### ISSUE TITLE
[Question] Current expectations of ram usage

### ISSUE THREAD

**Issue by bufke** [jedisct1/libsodium.js#128]
Hello, I noticed libsodium.js adds a 80mb of ram ArrayBuffer to my project. This seems to affect Wire too @bennyn - I was looking to see if it was just me. Reading #47 and the sumo docs I realize the pw_hash function does require a lot of ram. However is there any way to reduce this when not in use or set it dynamically? Some use cases I imagine would require a pw_hash on first load for logging in - and then not require it for a very long time (especially is users can remain logged in).

My tests were done just with the Firefox dev tools memory tab.

One idea I had that wouldn't require messing with libsodium.js would be to just move the entire sumo module to a web worker and then destroy the web worker when not in use. Memory usage would be even higher during login - but then much lower the rest of the time. I wanted to ask here first if there was anything obvious I'm missing.

**Comment by buu700** [comment]
This general issue just came up recently at https://github.com/jedisct1/libsodium/issues/631 and https://github.com/kripken/emscripten/issues/5925.

tl;dr: This was always possible with emscripten's `ALLOW_MEMORY_GROWTH` option, but only recently got become performant enough to be worth considering. In theory, that is; @jedisct's benchmarks on Node 9.x proved otherwise, which is now being investigated in that emscripten thread (possibly a V8 bug).

**Comment by buu700** [comment]
(Although, IIRC dynamically *decreasing* memory use is planned for a future version of wasm, in which case an approach like the one you described for your use case sounds reasonable to me.)

**Comment by bufke** [comment]
I wish I had some familiarity with Emscripten. My consulting company would be happy to fund some development if anyone had the interest.

@jedisct1 have you ever looked into https://www.coreinfrastructure.org/grants/ I wonder if libsodium could apply. I'd be happy to reach out and see if libsodium is the type of project they'd consider a grant for.

**Comment by bufke** [comment]
For anyone interested this is how you can manually change ram usage if you don't like the default. In my case I want the non sumo version but with a little more ram. If ram runs out - it will give an error like "C.print is not a function" and have enlargeMemory in the stack trace. It fails because it attempts to enlargeMemory but it doesn't work.

Defaults
(non-sumo) 16777216
(sumo) 83886080

You need to build libsodium.js which at this time is a little hard because the current version of emscripten doesn't work due to https://github.com/kripken/emscripten/issues/6387

This is tested in Ubuntu. First install the normal dependencies but install emscripten 1.37.35 (both older and newer version may not work)

- Install libtool and build-essential `apt install libtool build-essential`
- Install emscripten 1.37.35
  - `git clone https://github.com/juj/emsdk`
  - `./emsdk install clang-e1.37.35-64bit emscripten-1.37.35`
  - `./emsdk activate clang-e1.37.35-64bit emscripten-1.37.35`
  - `source emsdk_env.sh`
- Clone the libsodium.js repo and cd to it `git clone https://github.com/jedisct1/libsodium.js`
- Compile (Unsure if this order makes sense but it works for me) `make`
- `cd libsodium`
- `./configure`
- `make`
- To change ram, edit dist-build/emscripten.sh and look for where it says TOTAL_MEMORY. For example I made mine `TOTAL_MEMORY=50331648` to support pw_hash. Not all values are supported, I'm unsure what works and doesn't work. Doubling or tripling the existing value works. If using sumo edit the sumo version instead.
- `dist-build/emscripten.sh --standard` (make sure to run from this directory and not inside of dist-build, use --sumo instead of --standard if needed)
- Go back to libsodium.js directory `cd ..`
- `make`

The files should now be in `dist`.
